### ISSUE TITLE
Classifying Meanings & Documents - Fundamentals

### ISSUE THREAD

**Issue by HyunkuKwon** [Computational-Content-Analysis-2020/Readings-Responses-Spring#15]
Post questions here for one or more of our fundamentals readings:

Manning, Christopher, Prabhakar Raghavan and Hinrich Schütze.  2008. “[Text Classification and Naïve Bayes](https://nlp.stanford.edu/IR-book/pdf/13bayes.pdf)”, “[Vector Space Classification](https://nlp.stanford.edu/IR-book/pdf/14vcat.pdf),” and “[Support Vector Machines](https://nlp.stanford.edu/IR-book/pdf/15svm.pdf).” Chapters 13-16 from Introduction to Information Retrieval: 234-320.

or,

Witten, Ian H., Eibe Frank, Mark A. Hall, Christopher J. Pal. 2017. “[Ensemble Learning](https://github.com/Computational-Content-Analysis-2020/Readings-Responses/files/4052239/Witten2011.1.pdf)” Chapter 12 from Data Mining: Practical Machine Learning Tools and Techniques, 4th Edition: 351-371.

**Comment by Yilun0221** [comment]
I think ensemble is a very useful algorithm which enables us to combine the advantages of different algorithms. However, it is very hard to determine the hyperparameters and weight in the comprehensive algorithm of each algorithm. My intuition is that it depends on the spefic datasets and its fields. Practice and expertise in a certain kind of datasets may help us pick the optimal group of hyperparameters and the weights. I wonder if there is a general formula as a guidance for us to pick the parameters?

**Comment by wanitchayap** [comment]
The idea of stacking is very fascinating to me. I wonder if using a neural net type of learner for the level-1 model would work. That is, maybe we can concatenate the actual training data with output from level-0 models and feed the whole concatenated feature to the level-1 model (as often done in NLP tasks where lower and higher levels inputs are concatenated together). This way, the level-1 model may be able to learn more complicated rules? Also from my limited experiences with statistics and machine learnings, I don't really see the concept of stacking being used in research or being taught in stat classes. Is this technique actually used and developed in the recent years?

**Comment by ihsiehchi** [comment]
**Combining boosting and stacking?** Can we perform boosting, but instead of using models from the same function class (e.g. all decision trees), we combine diverse models as one would when performing stacking? That is what I thought boosting was doing when I read in the boosting section that for best results, it should be the case that "models are significantly different from one another." 

One immediate follow-up question would be: is there a conventional/optimal way of choosing the model classes (question applicable to just stacking too)? And is there an optimal order to iterate through those classes, possibly conditional upon the observed pattern of error rate?

**Comment by WMhYang** [comment]
**For Manning, Christopher, Prabhakar Raghavan and Hinrich Schütze. 2008.**

1. Chapter 14 - Vector Space Classification. In the bias-variance tradeoff part, it is proved that the learning error (the expectation of MSE) can be decomposed into bias part and variance part. Hence, minimizing MSE could be viewed as minimizing both bias and variance. I wonder whether it would be feasible to allocate different weights (instead of equal weights) to bias part and variance part in real applications so that we can minimize a constructed "quasi-MSE". If so, how could we determine the weight critieria?

2. Chapter 15- Support Vector Machine. On page 322, it was mentioned

> The functional margin of a data set with respect to a decision surface is then twice the functional margin of any of the points in the data set with minimal functional margin (the factor of 2 comes from measuring across the whole width of the margin, as in Figure 15.3).

I did not find a formal definition of "functional margin of a data set" in this chapter. What does that mean? Moreover, where does the relationship "twice" come from?

**Comment by frederickz** [comment]
I have some questions about the model discussed in Chapter 13 of _Introduction to Information Retrieval_. (I kind of like to get into details about the model.)

In the example of $c=China$, I am curious about why $B=6$.

**Comment by timqzhang** [comment]
**For "Text Classification and Naïve Bayes"**

When comparing the two models, namely the multinomial and Bernoulli model, the author mentions that the former one could handle longer docs while the latter one works best with short texts. I wonder whether there is a clear cutting point between the length of the docs when deciding which model to use? Also at p.272, it is mentioned that "weaker models are often preferable when limited training data are available". What does the "weaker" mean here, the less accuracy, or less risk to go overfit? Moreover, does it mean that when the data are quite limited in the training data, weaker models tend to face less overfit problem?

**For "Ensemble Learning"**

It is an inspiring chapter introducing many useful methods to realize the ensemble learning. Actually I have the similar concern as in the question above, which is that many methods introduced here do not mention the thresholds when we apply them. For example, in the "bagging", how many models of the same type should be in the ensemble? Or in the "stacking", what are the criteria to choose which model to be included? Further assessment could be done and general criteria could be provided, then we would have the clear reference when applying these method into the analyses.

**Comment by nwrim** [comment]
As demonstrated in the Netflix challenge, I think ensemble methods, especially ones that use different models (like the stacking method the Witten et al. article discusses) could be one of the most powerful methods in increasing the performance of models. I think the main takeaway from the Netflix challenge was that the "more the merrier" principle applies to models with relatively low performance - incorporating even the weaker model actually increase, or at least does not hurt the final model's performance. In a more computer-science space, this kind of massive ensemble makes sense since the performance is the criterion there, but in a more social-science space, I think this kind of method is less prevalent because of the interpretability problem. I have seen a lot of applications of Random Forest or Boost algorithm in social science, but I do not think I have yet seen a lot of cases where a bunch of different models is combined together in the aspect of social science. Could you give us some illustrative example of an ensemble approach of multiple distinct methods in social science, regardless of whether it is from your own research experience or not?

**Comment by linghui-wu** [comment]
In the k-nearest neighbor part of *Vector Space Classification*, the author mentions that:

> The parameter *k* in KNN is often chosen based on experience or knowledge about the classification problem at hand. It is desirable for k to be odd to make ties less likely. *k*=3 and *k*=5 are common choices, but much larger values between 50 and 100 are also used.

I would like to know if there is theoretical or practical evidence that *k*=3 or *k*=5 would be better choices. Or is there any rule of thumb that helps us in picking the optimal parameters? Does the parameter related to the number of target categories such as the topics in a given text?

As for @timqzhang's comment on *Text Classification and Naive Bayes*, I think the "weaker" models are useful in avoiding overfitting. According to my limited knowledge about the machine learning, I guess another advantage of a "weaker" models on limited training data is that they have fewer strict presumptions, which might be beneficial in modeling. Someone, please correct me if I am wrong! : )

**Comment by tianyueniu** [comment]
Based on past experiences in dealing with ML models, I have a general question on comparing models and choosing the best one. I have encountered scenarios where I ran KNN, Naive Bayes and tree-based models on the same data and achieved similar performances. Sometimes different iterations might even return different 'best model', even with cross-validation. In that case, how should we choose between different models? How should we balance accuracy, variances, complexity and computational efficiency?

**Comment by minminfly68** [comment]
For Manning, Christopher, Prabhakar Raghavan and Hinrich Schütze. 2008.

It has a very detailed introduction to Naive Bayes (NB) classification. I am just wondering why author claimed that "weaker models are often preferable when limited training data are available?"? If more hands-on experience could be provided for using these classifiers for textual documents would be really appreciated. Also, if you could explain more about when to apply weaker ML models, it would also be very helpful.

**Comment by DSharm** [comment]
Echoing some of the questions above about best practices and tuning hyperparameters - whats a rule of thumb for how large the training dataset should be for ensemble methods to be applied?

**Comment by harryx113** [comment]
Adding on to Nak Won's comment, I am wondering the possible downsides of ensemble methods besides interpretability. We know ML is not a "one size fits all" problem, but would ensemble lead to overfitting in some way?

**Comment by Lesopil** [comment]
So I am wondering about the field a bit more broadly again. Many of these techniques require a great deal of computing power. I wonder if we are waiting for further developments in the power of computers and processors to make more/better conclusions, or of the main leaps forward come in the form of new algorithms/models?

**Comment by jsgenan** [comment]
I was a bit surprised to see MultinomialNB has the best performance in the code notebook. Can we talk a little bit about what's the comparative advantage of each method?

**Comment by liu431** [comment]
I am interested in the rotation forests as I have never seen it before. Although the chapter describes its advantage of getting diversified and accurate results, it might be less interpretable than random forests. In a random forest, we can still get the feature importance scores. But in the rotation forests, the principal components are constructed on different subsets. Thus, is this one of its disadvantages if we also care about the importance of individual features?

**Comment by pdiazm** [comment]
My question on these papers is about the trade-off between accuracy and interpretability. If our objective is to design an intervention to regulate or mediate content, it could be very useful to understand how the algorithm is making decisions - especially if we want to intertwine human supervision into our mediation process. How can we recover some of the interpretability of these methods?

**Comment by bazirou** [comment]
Before text classification, we have to pre-process the data. So I wonder what the general process is.

**Comment by lyl010** [comment]
It is very helpful to notice that the historical path of development of NLP: naive Bayes, SVD, and vector representation. I have always been fascinated  by the Word 2 Vec and then I found that W2V is an extension from measuring the probability of words, so it does have theory and not as the black box as we thought before.
My question is somehow a general one, do you think learning the theories of method can be helpful for future computational social scientists?
